core/ordersync/ordersync.go

```diff
@@ -312,65 +311,60 @@ func (s *Service) GetOrders(ctx context.Context, minPeers int) error {
 				"provider": peerID.Pretty(),
 			}).Trace("requesting orders from neighbor via ordersync")
 
+			select {
+			case <-innerCtx.Done():
+				break
+			case semaphore <- struct{}{}:
+			}
+
 			wg.Add(1)
 			go func(id peer.ID) {
-				defer wg.Done()
-				select {
-				case <-innerCtx.Done():
-					// NOTE(jalextowle): In this case, we haven't written to the semaphore
-					// so we shouldn't read from it.
-					return
-				case semaphore <- struct{}{}:
-					defer func() { <-semaphore }()
-				}
+				defer func() {
+					wg.Done()
+					<-semaphore
+				}()
 				if err := s.getOrdersFromPeer(innerCtx, id); err != nil {
 					log.WithFields(log.Fields{
 						"error":    err.Error(),
 						"provider": id.Pretty(),
 					}).Warn("could not get orders from peer via ordersync")
 				} else {
-					// TODO(albrow): Handle case where no orders were returned from this
```

Comment:
I just decided to treat a response of 0 as an error in my implementation of `makeOrderSyncRequest`. @albrow I'd be interested to hear your thoughts on whether or not this appropriate. I didn't think it was necessary personally, but I'd be open to logging a different message in the if-statement above.